aten/src/ATen/native/cuda/SegmentReduce.cu

```diff
@@ -97,54 +269,78 @@ Tensor _segment_reduce_cuda_kernel(
         auto* data_data_ptr = data.data_ptr<scalar_t>();
         auto* output_data_ptr = output.data_ptr<scalar_t>();
 
-        if (reduction == SegmentReductionType::MAX) {
-          CustomMax max_op{};
-          scalar_t initial_value = initial.has_value()
-              ? initial.value().to<scalar_t>()
-              : std::numeric_limits<scalar_t>::lowest();
-          CUB_WRAPPER(
-              cub::DeviceSegmentedReduce::Reduce,
-              data_data_ptr,
-              output_data_ptr,
-              segment_count,
-              offsets_data_ptr,
-              offsets_data_ptr + 1,
-              max_op,
-              initial_value,
-              at::cuda::getCurrentCUDAStream());
+        // initialize starting value
+        scalar_t initial_value;
+        if (initial.has_value()) {
+          initial_value = initial.value().to<scalar_t>();
+        } else if (reduction == SegmentReductionType::MAX) {
+          initial_value = std::numeric_limits<scalar_t>::lowest();
         } else if (reduction == SegmentReductionType::MEAN) {
-          CustomSum sum_op{};
-          scalar_t initial_value = initial.has_value()
-              ? initial.value().to<scalar_t>()
-              : (scalar_t)0;
-          CUB_WRAPPER(
-              cub::DeviceSegmentedReduce::Reduce,
-              data_data_ptr,
-              output_data_ptr,
-              segment_count,
-              offsets_data_ptr,
-              offsets_data_ptr + 1,
-              sum_op,
-              initial_value,
-              at::cuda::getCurrentCUDAStream());
-
-          post_sum_div_kernel<scalar_t>
+          initial_value = 0;
+        }
+
+        if (output_shape.size() > 1) {
```

Comment:
if input/output is multi-d, but the non-reduced dimensions are 1 (input.size(0) == input.numel()) you are better off taking CUB path, but that's an optimization for later.